pkg/server/server.go

```diff
@@ -238,11 +285,24 @@ func (s *ProxyServer) Proxy(stream client.ProxyService_ProxyServer) error {
 		for {
 			in, err := stream.Recv()
 			if err == io.EOF {
+				klog.V(5).Infoln("Received EOF from stream")
 				close(stopCh)
 				return
 			}
 			if err != nil {
-				klog.ErrorS(err, "Stream read from frontend failure")
+				if agentID, lookupErr := frontStream.getAgentID(); lookupErr == nil {
+					if frontStream.getStatus() != Closing {
+						klog.ErrorS(err, "Stream read from frontend failure",
+							"agentID", agentID)
+						s.BackendManager.TaintBackend(agentID, err)
```

Comment:
Why do we taint the backend when the frontend stream read is broken?

Reply:
Sadly if the lose the backend to something like a machine crash or network outage we get no error on the backend. Essentially we send TCP packets which disappear into the ether. TCP timeout or even keepalive is generally in minutes and the frontend request timeout is in seconds. So the error shows up as a front ends error (really a timeout but to us it looks like a context closed error). We can't be sure what is wrong, so its better to "taint" the connection than close it. We need to rely on the keepalive options others have added to take care of that. However keepalive can be slow. This helps us to not send requests to a dead connection while we are determining if the connection is actually dead.

Reply:
Does this mean hitting any client side timeout results in a backend taint? I've encountered timeouts on the k/k client side and sometimes it's just the destination service being overloaded. This happens a decent amount in the k/k e2e tests and the retry mechanism for most clients takes care of the problem.

I'm worried a problematic destination (eg: webhook) could cause all backends to be tainted for a short period of time.

Reply:
All backends getting tainted just reverts us to the current behavior. We lose the benefits of tainting but the system continues to work.

Reply:
We can detect broken agent by observing if a dial response is returned by the agent on time. That's less prone to false-positive.